File: articles/cell-theory-basic-unit-of-life/r/treatment_response.R
# Treatment-response summary in R.

counts_path <- file.path("data", "cell_counts.csv")
viability_path <- file.path("data", "viability_observations.csv")

if (!file.exists(counts_path)) {
  counts_path <- file.path("..", "data", "cell_counts.csv")
  viability_path <- file.path("..", "data", "viability_observations.csv")
}

counts <- read.csv(counts_path)
viability <- read.csv(viability_path)

growth_summary <- do.call(
  rbind,
  lapply(split(counts, counts$condition), function(df) {
    fit <- lm(log(cells) ~ time_h, data = df)
    r <- coef(fit)[["time_h"]]

    data.frame(
      condition = unique(df$condition),
      growth_rate_per_h = r,
      doubling_time_h = log(2) / r
    )
  })
)

viability_summary <- do.call(
  rbind,
  lapply(split(viability, viability$condition), function(df) {
    fit <- lm(log(viable_cells) ~ time_h, data = df)
    k <- -coef(fit)[["time_h"]]

    data.frame(
      condition = unique(df$condition),
      loss_rate_per_h = k,
      viability_half_life_h = log(2) / k
    )
  })
)

print(round(growth_summary, 5))
print(round(viability_summary, 5))

File: articles/cell-theory-basic-unit-of-life/python/treatment_response.py
"""
Treatment-response summary workflow.

Run:
    python python/treatment_response.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from cell_theory_core import fit_exponential_growth, fit_viability_decay


ARTICLE_DIR = Path(__file__).resolve().parents[1]
COUNTS_PATH = ARTICLE_DIR / "data" / "cell_counts.csv"
VIABILITY_PATH = ARTICLE_DIR / "data" / "viability_observations.csv"


def main() -> None:
    """Compare growth and viability parameters by condition."""

    counts = pd.read_csv(COUNTS_PATH)
    viability = pd.read_csv(VIABILITY_PATH)

    growth_rows = []

    for condition, group in counts.groupby("condition"):
        fit = fit_exponential_growth(
            group["time_h"].to_numpy(dtype=float),
            group["cells"].to_numpy(dtype=float),
        )

        growth_rows.append(
            {
                "condition": condition,
                "growth_rate_per_h": fit.growth_rate_per_h,
                "doubling_time_h": fit.doubling_time_h,
            }
        )

    decay_rows = []

    for condition, group in viability.groupby("condition"):
        fit = fit_viability_decay(
            group["time_h"].to_numpy(dtype=float),
            group["viable_cells"].to_numpy(dtype=float),
        )

        decay_rows.append(
            {
                "condition": condition,
                "loss_rate_per_h": fit.loss_rate_per_h,
                "viability_half_life_h": fit.half_life_h,
            }
        )

    growth_df = pd.DataFrame(growth_rows)
    decay_df = pd.DataFrame(decay_rows)

    print(growth_df.round(5).to_string(index=False))
    print(decay_df.round(5).to_string(index=False))


if __name__ == "__main__":
    main()
